only allow adding unused columns

machisuji opened the pull request:
Work in Progress

WP [#19553](https://community.openproject.org/work_packages/19553)

This probably warrants a test. I will need at least another day for that, though.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://machisuji-fix-duplicate-available-columns.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-fix-duplicate-available-columns.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

machisuji commented:
This breaks rendering of added columns. I don't see the connection yet as I have no idea when it comes to anything in the frontend, specifically Angular. Also debugging angular is no fun.

![screen shot 2015-04-13 at 17 29 49](https://cloud.githubusercontent.com/assets/158871/7120126/b14501f6-e202-11e4-8872-23b2cae45796.png)

The screenshot shows on the right the added 'Project' column whose heading and values cannot be rendered due to the errors seen on in the debugger.

machisuji commented:
It does fix the column selection, though! :D

floriank commented on frontend/app/services/query-service.js:
```diff
@@ -164,10 +164,11 @@ module.exports = function(Query, Sortation, $http, PathHelper, $q, AVAILABLE_WOR
-          availableUnusedColumns = WorkPackagesTableHelper.getColumnDifference(available_columns, QueryService.getSelectedColumns());
-          return availableUnusedColumns;
-        });
+        .then(QueryService.selectUnusedColumns);
```
if this is only used within the service, it should not be exposed in the service. could you just add the function to the outer scope, even outside the `QueryService` definition?

NobodysNightmare commented:
@machisuji Are you still working on this? Are you road blocked here?

machisuji committed:
only unused columns should be available as new columns

machisuji commented:
Stepped away from the problem for a bit and coming back to it I see it clearly. I didn't see or consider the [side effects](https://github.com/opf/openproject/blob/dev/frontend/app/services/query-service.js#L168). Anyway here we go. This fixes it for real this time.

NobodysNightmare commented on frontend/app/work_packages/controllers/dialogs/columns.js:
```diff
@@ -41,7 +41,7 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
   vm.promise = QueryService.loadAvailableColumns()
     .then(function(availableColumns){
-      vm.availableColumns = availableColumns;
+      vm.availableColumns = QueryService.getAvailableUnusedColumns();
```
Still have to ask:

The code reads: 
1. load available columns
2. then disregard those columns (given as parameter)
3. and take available unused columns instead

I guess the answer will be something like:
`loadAvailableColumns` does the http request, while `getAvailableUnusedColumns` does no networking, but returns the right thing...

Still it looks weird, to have this single most return value of `loadAvailableColumns` ignored...

machisuji replied:
It's not ignored. Read on and [you will see](https://github.com/machisuji/openproject/blob/fix/duplicate_available_columns/frontend/app/work_packages/controllers/dialogs/columns.js#L45) that it's actually still used.

NobodysNightmare commented:
Arg...

Sorry @machisuji but this is not the fix.

Yes, I can't add e.g. the ID column if it is already added.
But after reomving a column from the selection, I still can't add it again!

### Reproduction
1. open the column selection
2. remove any column that is currently included
3. try to re-add that column

machisuji committed:
update unused columns upon change of selection

machisuji commented:
@NobodysNightmare Fixed the issue you mentioned. The problem was that the columns available for selection were only updated when the modal was opened, not when the selection changes.

@myabc Unfortunately I implemented this with the refresh attribute of the select element. This is called with every key stroke the user makes when searching for a column. Ideally this would only happen when the selection changes (column removed, added), though. I couldn't get that to work, though, unfortunately (tried `ng-update` and `update` to no avail). Do you know how to do this?

All I need to happen is for the function `#updateUnusedColumns(selectedColumns)` to be called every time the selection (i.e. a column is added or removed from the selection) changes.

machisuji committed:
fixed naming and line length

NobodysNightmare commented on frontend/app/work_packages/controllers/dialogs/columns.js:
```diff
@@ -35,13 +35,16 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
   vm.selectedColumns = [];
   vm.oldSelectedColumns = [];
   vm.availableColumns = [];
+  vm.unusedColumns = [];
```
What is the semantic difference between an available column and an unused column? Are both lists equal?

machisuji replied:
Available columns are all columns you could ever select. Unused columns are those that are not currently selected.

So basically: `unusedColumns = availableColumns - selectedColumns`

NobodysNightmare commented on frontend/app/work_packages/controllers/dialogs/columns.js:
```diff
@@ -55,10 +58,12 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
+    return column.name;
+  }
+
   function getColumnNames(arr) {
```
Arrrrrr!

![Arrrr!](http://typophile.com/files/Arrrrr2_6628.gif)

machisuji commented:
While the trick with the refresh is not the prettiest it is a valid solution and it works just fine. So unless I find out how to get notified of choice updates we can just use this IMO.

NobodysNightmare commented:
Looks like a valid solution... If someone dislikes it, then he/she will have to make a new PR ;-P

floriank commented on frontend/app/work_packages/controllers/dialogs/columns.js:
```diff
@@ -55,10 +58,12 @@ module.exports = function($scope, $filter, columnsModal, QueryService, WorkPacka
     return _.difference(vm.selectedColumns, vm.oldSelectedColumns);
   }
 
+  function getColumnName(column) {
```
I do not get the Mehrwert of this method.

machisuji replied:
It saved me from writing the same inline function - which looks horrible in overly verbose Javascript - twice.
Probably saved me effectively a line while making the code more readable IMO.
